webui/react/src/stores/agents.tsx

```diff
@@ -0,0 +1,100 @@
+import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
+
+import { getAgents } from 'services/api';
+import { clone } from 'shared/utils/data';
+import { percent } from 'shared/utils/number';
+import { noOp } from 'shared/utils/service';
+import { Agent, ClusterOverview, ClusterOverviewResource, ResourceType } from 'types';
+import handleError from 'utils/error';
+import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+
+type AgentContext = {
+  agents: Loadable<Agent[]>;
+  updateAgents: (a: Loadable<Agent[]>) => void;
+};
+
+const initResourceTally: ClusterOverviewResource = { allocation: 0, available: 0, total: 0 };
+// TODO: dont export
+export const initClusterOverview: ClusterOverview = {
+  [ResourceType.CPU]: clone(initResourceTally),
+  [ResourceType.CUDA]: clone(initResourceTally),
+  [ResourceType.ROCM]: clone(initResourceTally),
+  [ResourceType.ALL]: clone(initResourceTally),
+  [ResourceType.UNSPECIFIED]: clone(initResourceTally),
+};
+
+const AgentsContext = createContext<AgentContext>({ agents: NotLoaded, updateAgents: noOp });
+
+export const AgentsProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const [state, setState] = useState<Loadable<Agent[]>>(NotLoaded);
+  return (
+    <AgentsContext.Provider value={{ agents: state, updateAgents: setState }}>
+      {children}
+    </AgentsContext.Provider>
+  );
+};
+
+export const useFetchAgents = (canceler: AbortController): (() => Promise<void>) => {
+  const { updateAgents } = useContext(AgentsContext);
+
+  return useCallback(async (): Promise<void> => {
+    try {
+      const agents = await getAgents({ signal: canceler.signal });
+      updateAgents(Loaded(agents));
+    } catch (e) {
+      handleError(e);
+    }
+  }, [canceler, updateAgents]);
+};
+
+export const useEnsureAgentsFetched = (canceler: AbortController): (() => Promise<void>) => {
+  const { agents, updateAgents } = useContext(AgentsContext);
+
+  return useCallback(async (): Promise<void> => {
+    if (agents !== NotLoaded) return;
+    try {
+      const agents = await getAgents({ signal: canceler.signal });
+      updateAgents(Loaded(agents));
+    } catch (e) {
+      handleError(e);
+    }
+  }, [canceler, updateAgents, agents]);
+};
```

Comment:
Your list is accurate but in my experience you want to give components control over the freshness of data. In this particular case you don't need to but I want the stores to have parity of functionality so it's easier to work with them.

Reply:
(Feel free to ignore this one if it is getting too far afield for this PR)

what about something like `useAgents` and `useFreshAgents`?  (okay the name is pretty bad).

at that point the functionality would be basically the same as existing `useFetchAgents`/`useEnsureAgentsFetched` (i think)

but then if the calls happen within the provider itself it could be possible to make it so that the fetch only happens once even if `useFreshAgents` is happening in multiple places.

Reply:
So the issue there is you're talking about the "getting" side which shouldn't know about how the data arrives. Neither `useAgents` nor `useFreshAgents` will do anything unless an appropriate action is triggered somewhere else.